provider/adapters/src/fallback/adapter.rs

```diff
@@ -3,25 +3,186 @@
+{
+    /// Create a [`LocaleFallbackProvider`] by wrapping another data provider and then loading
+    /// fallback data from it.
+    ///
+    /// If the data provider being wrapped does not contain fallback data, use
+    /// [`LocaleFallbackProvider::new_with_fallbacker`].
+    pub fn try_new(provider: P) -> Result<Self, DataError> {
+        let fallbacker = LocaleFallbacker::try_new(&provider)?;
+        Ok(Self {
+            inner: provider,
+            fallbacker,
+        })
+    }
 }
 
-pub struct LocaleFallbackAdapter<P>(pub P);
+impl<P> LocaleFallbackProvider<P> {
+    /// Wrap a provider with an arbitrary fallback engine.
+    ///
+    /// This relaxes the requirement that the wrapped provider contains its own fallback data.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use icu_locid::locale;
+    /// use icu_provider::hello_world::*;
+    /// use icu_provider::prelude::*;
+    /// use icu_provider_adapters::fallback::{LocaleFallbacker, LocaleFallbackProvider};
+    ///
+    /// let provider = HelloWorldProvider::new_with_placeholder_data();
+    ///
+    /// let req = DataRequest {
+    ///     options: locale!("de-CH").into(),
+    ///     metadata: Default::default(),
+    /// };
+    ///
+    /// // There is no "de-CH" data in the `HelloWorldProvider`
+    /// ResourceProvider::<HelloWorldV1Marker>::load_resource(&provider, &req)
+    ///     .expect_err("No data for de-CH");
+    ///
+    /// // `HelloWorldProvider` does not contain fallback data,
+    /// // but we can fetch it from `icu_testdata`, and then
+    /// // use it to create the fallbacking data provider.
+    /// let fallbacker = LocaleFallbacker::try_new(&icu_testdata::get_provider())
+    ///     .expect("Fallback data present");
+    /// let provider = LocaleFallbackProvider::new_with_fallbacker(provider, fallbacker);
+    ///
+    /// // Now we can load the "de-CH" data via fallback to "de".
+    /// let german_hello_world: DataPayload<HelloWorldV1Marker> = provider
+    ///     .load_resource(&req)
+    ///     .expect("Loading should succeed")
+    ///     .take_payload()
+    ///     .expect("Data should be present");
+    ///
+    /// assert_eq!("Hallo Welt", german_hello_world.get().message);
+    /// ```
+    pub fn new_with_fallbacker(provider: P, fallbacker: LocaleFallbacker) -> Self {
+        Self {
+            inner: provider,
+            fallbacker,
+        }
+    }
+}
+
+impl<P> AnyProvider for LocaleFallbackProvider<P>
+where
+    P: AnyProvider,
+{
+    fn load_any(&self, key: ResourceKey, base_req: &DataRequest) -> Result<AnyResponse, DataError> {
+        let key_fallbacker = self.fallbacker.for_key(key);
+        let mut fallback_iterator = key_fallbacker.fallback_for(base_req.clone());
+        while !fallback_iterator.get().options.is_empty() {
+            let result = self.inner.load_any(key, fallback_iterator.get());
+            if !result_is_err_missing_resource_options(&result) {
+                return result;
```

Comment:
Do we actually want to return data errors with the `fallback_iterator.get()` options set, or do we want to hide that from the client?

Reply:
I'm catching `MissingResourceOptions` because that's the soft error that we can recover from. If the error is something else more severe, like a Serde deserialization error or a type mismatch error, then we should propagate it. In languages with `try/catch` statements, you only `catch` the error you want to handle, and that's what I'm doing here

Reply:
I understand. But any error from the underlying provider *that you don't catch* will have the modified request as it's context. This might be confusing, so I propose that you also add the original request:
```rust
return result.map_err(|e| e.with_req(key, &base_req));
```

Reply:
Done.